dubbo-remoting/dubbo-remoting-http12/src/main/java/org/apache/dubbo/remoting/http12/h2/Http2ServerChannelObserver.java

```diff
@@ -74,12 +76,33 @@
 
     @Override
     public void cancel(Throwable throwable) {
+        if (throwable instanceof CancelStreamException) {
+            if (((CancelStreamException) throwable).isCancelByRemote()) {
+                closed = true;
+            }
+        }
+        this.cancellationContext.cancel(throwable);
         long errorCode = 0;
         if (throwable instanceof ErrorCodeHolder) {
             errorCode = ((ErrorCodeHolder) throwable).getErrorCode();
         }
         getHttpChannel().writeResetFrame(errorCode);
-        this.cancellationContext.cancel(throwable);
+    }
+
+    @Override
+    public void onNext(Object data) {
+        if (closed) {
+            return;
+        }
+        super.onNext(data);
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        if (closed) {
+            return;
+        }
+        super.onError(throwable);
```

Comment:
**Optimization Issue**: Similar to onNext, the onError method returns early if the channel is closed, potentially missing important error handling steps. This could lead to unhandled exceptions or errors that occur after the channel is marked as closed but before it is actually closed. <br> **Fix**: Ensure that errors occurring after the channel is marked as closed but before it is actually closed are properly logged or handled to prevent unhandled exceptions. <br> **Code Suggestion**: 
 ```
 + if (closed) {
 +    logError(throwable); // Implement this method to log or handle error
 +    return;
 + }
 ```